### PR TITLE
feat: Balancer V2 integration

### PR DESCRIPTION
**Pull request by xianny**
## Description

Liquidity is severely lacking at the moment but did a very limited simulation prod vs **Balancer V2 only**
**Sells**
image
**Buys**
image

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Review by dorothy-zbornak — COMMENTED** [review]
Still reviewing the TS but liking what I see!

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        IERC20TokenV06 sellToken,
+        IERC20TokenV06 buyToken,
+        uint256 sellAmount,
+        bytes memory bridgeData
+    )
+        internal
+        returns (uint256 boughtAmount)
+    {
+        // Decode the bridge data.
+        (IBalancerV2Vault vault, bytes32 poolId, uint256 deadline) = abi.decode(
```
[the remark this answers is not in the document]
Comment (kimpers) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol:89]: Not sure if passing the deadline brings any value here, it just diverges from other sources. Will update to `block.timestamp` 👍

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/cream_utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    protected async _fetchPoolsForPairAsync(takerToken: string, makerToken: string): Promise<Pool[]> {
+        try {
+            const poolData = (await getPoolsWithTokens(takerToken, makerToken)).pools;
+            // Sort by maker token balance (descending)
+            const pools = parsePoolData(poolData, takerToken, makerToken).sort((a, b) =>
+                b.balanceOut.minus(a.balanceOut).toNumber(),
+            );
+            return pools.length > this.maxPoolsFetched ? pools.slice(0, this.maxPoolsFetched) : pools;
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/cream_utils.ts:20]:
```suggestion
            return pools.slice(0, this.maxPoolsFetched);
```

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    //         return pools.length > this.maxPoolsFetched ? pools.slice(0, this.maxPoolsFetched) : pools;
+    //     } catch (err) {
+    //         return [];
+    //     }
+    // }
+
+    protected async _fetchPoolsForPairAsync(takerToken: string, makerToken: string): Promise<Pool[]> {
+        const query = `
+        query getPools {
+            pools(where: {
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/balancer_v2_utils.ts:35]: do we also want a `first: ${this._maxPoolsFetched}`?

File: packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pools_cache.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// tslint:disable:custom-no-magic-numbers
+const FIVE_SECONDS_MS = 5 * 1000;
+const ONE_DAY_MS = 24 * 60 * 60 * 1000;
+const DEFAULT_TIMEOUT_MS = 1000;
+// tslint:enable:custom-no-magic-numbers
+
+export abstract class PoolsCache {
+    constructor(protected readonly _cache: { [key: string]: CacheValue }, protected readonly maxPoolsFetched: number) {}
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/utils/market_operation_utils/pools_cache/pools_cache.ts:15]: Why does `maxPoolsFetched` belong to this class? It doesn't use it.

**Review by dorothy-zbornak — APPROVED** [review]
really nice work on this!
:+1: :eye: :lips: :eye: :+1:

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 5]
```diff
@@ -100,29 +100,16 @@ export class MarketOperationUtils {
...
-        const {
-            onChain: sampleBalancerOnChain,
-            offChain: sampleBalancerOffChain,
-        } = this._sampler.balancerPoolsCache.howToSampleBalancer(
-            takerToken,
-            makerToken,
-            quoteSourceFilters.isAllowed(ERC20BridgeSource.Balancer),
+        // Can't sample Balancer or Cream on-chain without the pools cache
+        const sourcesWithCaches = [ERC20BridgeSource.BalancerV2, ERC20BridgeSource.Balancer, ERC20BridgeSource.Cream];
+        const excludeSources: ERC20BridgeSource[] = sourcesWithCaches.filter(
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:105]: `staleCachedSources`?

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts [hunk 6]
```diff
@@ -88,6 +90,7 @@ export const BATCH_SOURCE_FILTERS = SourceFilters.all().exclude([ERC20BridgeSour
  */
 export class SamplerOperations {
     public readonly liquidityProviderRegistry: LiquidityProviderRegistry;
+    public readonly poolsCaches: { [key in ERC20BridgeSource]?: PoolsCache };
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:93]: maybe worth defining `type SourcesWithPoolsCache = ERC20BridgeSource.Balancer | ERC20BridgeSource.BalancerV2 | ERC20BridgeSource.Cream`

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 7]
```diff
@@ -179,9 +157,7 @@ export class MarketOperationUtils {
                 rawTwoHopQuotes,
                 isTxOriginContract,
             ],
-            offChainBalancerQuotes,
-            offChainCreamQuotes,
-        ] = await Promise.all([samplerPromise, offChainBalancerPromise, offChainCreamPromise]);
+        ] = await Promise.all([samplerPromise, cacheRefreshPromises]);
```
Comment (mzhu25) [inline packages/asset-swapper/src/utils/market_operation_utils/index.ts:160]: is this going to refresh the caches even on a BSC API request?

**Review by mzhu25 — COMMENTED** [review]
looking real good! Thanks for refactoring the pools cache stuff. Left a question about what happens on a BSC API request

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (xianny) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol]: Wasn't sure what `deadline` we want to use, just double-checking we definitely want it to be valid for only the current block? Won't that be a bit short?
https://github.com/balancer-labs/balancer-core-v2/blob/master/contracts/vault/interfaces/IVault.sol#L488-L489

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts (hunk 7 above)
Comment (xianny) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:160]: Yesss but I think that's what it did before. What's the desired behaviour?

Comment (mzhu25) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts:160]: I don't have a strong preference, but I lean towards not refreshing the caches for a BSC request. Should be fine either way though

**Review by mzhu25 — APPROVED** [review]
nice work!

File: contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinBalancerV2.sol]: You can just +1 it to be safe

File: packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts [hunk 8]
```diff
@@ -1187,29 +1156,56 @@ export class SamplerOperations {
...
-                                    this.getBalancerSellQuotes(
-                                        poolAddress,
-                                        makerToken,
-                                        takerToken,
-                                        takerFillAmounts,
-                                        ERC20BridgeSource.Cream,
-                                    ),
-                                );
+                            return (
+                                this.poolsCaches[ERC20BridgeSource.Cream].getCachedPoolAddressesForPair(
```
Comment (dekz) [inline packages/asset-swapper/src/utils/market_operation_utils/sampler_operations.ts:1196]:
```
case ERC20BridgeSource.Cream:
case ERC20BridgeSource.Balancer:
    this.poolsCaches[source].getCachedPoolAddressesForPair(...)
    this.getBalancerSellQuotes(
                                    poolAddress,
                                    makerToken,
                                    takerToken,
                                    takerFillAmounts,
                                    source);
```

**Review by dekz — APPROVED** [review]

**Comment by dekz** [comment]
Just recall we need to get this `FQT` and `BridgeAdapter` out before enabling `BalancerV2` in 0xAPI production. So be careful when you merge (incase another team publishes master and it rolls out before we're ready)

**Comment by dekz** [comment]
Contract changes have been deployed on mainnet https://github.com/0xProject/protocol/pull/225

**Comment by kimpers** [comment]
> IMO this promise should immediately resolve if the cache has been refresh recently.
> 
> I don't think we should be making a request to the graph for every API request we receive.

The promise array is built from the list of sources with stale cache so most of the time this will just be an empty array, but looking at the code I don't this works as expected 😂 . It's trying to do 
```typescript
await Promise.all([samplerPromise, [cachePromises]])
```
But I'm fairly certain that we only wait for `samplerPromise` and do nothing with `[cachePromises]`
It should be more like this
```typescript
await Promise.all([samplerPromise, Promise.all([cachePromises])])
```

I'll update and double check this

File: packages/asset-swapper/src/utils/market_operation_utils/index.ts [hunk 9]
```diff
@@ -147,7 +134,7 @@ export class MarketOperationUtils {
             ),
             // Get sell quotes for taker -> maker.
             this._sampler.getSellQuotes(
-                quoteSourceFilters.exclude(offChainSources).sources,
+                quoteSourceFilters.exclude(sourcesWithStaleCaches).sources,
```
[the remark this answers is not in the document]
Comment (kimpers) [reply packages/asset-swapper/src/utils/market_operation_utils/index.ts]: Updated to not exclude stale cache sources 👍

**Comment by kimpers** [comment]
Note: this should build after Uniswap V3 has been merged after it.

**Pull request merged**